### PR TITLE
`ByteBuffer` uses `ArrayPool` to prevent LOH allocations

### PR DESCRIPTION
**Pull request by TIHan**
loh
We allocate a lot on the LOH when serializing FSharp metadata. This PR should help.

### CODE REVIEW

**Review by cartermp — APPROVED** [review]
I like the approach, and it seems great to not constantly reallocate those large buffers.

File: src/fsharp/utils/FileSystem.fs [hunk 1]
```diff
@@ -786,52 +806,90 @@ type internal ByteBuffer =
...
+    member buf.Position =
+        buf.CheckDisposed()
+        buf.bbCurrent
 
-    static member Create sz =
-        { bbArray = Bytes.zeroCreate sz
+    static member Create(sz, useArrayPool) =
+        { useArrayPool = defaultArg useArrayPool false
+          isDisposed = false
+          bbArray = ArrayPool.Shared.Rent sz
```
Comment (baronfel) [inline src/fsharp/utils/FileSystem.fs:882]: This should depend on `useArrayPool` if I'm reading the intent right?

Comment (TIHan) [reply src/fsharp/utils/FileSystem.fs:896]: Yes, that is right

File: eng/Versions.props [hunk 2]
```diff
@@ -75,6 +75,7 @@
       https://pkgs.dev.azure.com/azure-public/vside/_packaging/vs-impl/nuget/v3/index.json;
     </RestoreSources>
     <!-- System.* packages -->
+    <SystemRuntimeCompilerServicesUnsafeVersion>4.7.1</SystemRuntimeCompilerServicesUnsafeVersion>
```
Comment (TIHan) [inline eng/Versions.props:78]: @KevinRansom you may not like this, but the changes I made with the ByteBuffer, is calling something from System.Runtime.CompilerServices.Unsafe, which has problems with binding redirects. Best way to resolve this is include the specific version.

**Review by KevinRansom — COMMENTED** [review]
This looks goods, I wonder why we have so many different sizes for allocation buffers, 100000, 50000, 20000 and 10000.  Meaningful names might help me understand why each size was chosen,  Of course it might not too.

File: src/fsharp/QuotationPickler.fs [hunk 3]
```diff
@@ -363,20 +369,26 @@ module SimplePickle =
...
-                { os = ByteBuffer.Create 100000
-                  ostrings=Table<_>.Create() }
             p x st1
-            st1.ostrings.AsList, st1.os.Close()
+            st1.ostrings.AsList, st1.os.AsMemory()
+
         let phase2data = (stringTab, phase1bytes)
+
+        let st2 =
+           { os = ByteBuffer.Create(100000, useArrayPool = true)
```
[the remark this answers is not in the document]
Comment (TIHan) [reply src/fsharp/QuotationPickler.fs:382]: I have no idea why these values were chosen. This was here before.

[the remark this answers is not in the document]
Comment (KevinRansom) [reply src/fsharp/QuotationPickler.fs:382]: Yeah, just give it a name.  pickle_object_buffersize or somesuch.

[the remark this answers is not in the document]
Comment (TIHan) [reply src/fsharp/QuotationPickler.fs:386]: I added the literals.

**Comment by TIHan** [comment]
Addresses this issue: https://github.com/dotnet/fsharp/issues/7982
Only took a year and half...

**Comment by cartermp** [comment]
note that azdo is down for us, so there's no point in trying to re-trigger CI

**Pull request merged**